### PR TITLE
Toreview

### PR DESCRIPTION
**Pull request by TheSpaceSheep**
plot_dataframes.py and display_interactions.py are for plotting purposes.

##new features : 
*masking by adding a new value dedicated to represent masking
*storing interactions during training so we can look at input-message pairs
*plotting using pandas+seaborn

### CODE REVIEW

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 1]
```diff
@@ -43,15 +55,15 @@ def __init__(self,
             mode: str,
             initial_n_unmasked: int=1
         ):
-        valid_modes = ['left_to_right', 'random']
+        valid_modes = ['left_to_right', 'random', 'dedicated_value']
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: It seems to me like dedicated value could be a separate flag that could apply to both left_to_right and random

File: egg/zoo/compo_vs_generalization/data.py [hunk 2]
```diff
@@ -216,6 +223,10 @@ def mask_attributes(sender_input, idxs_to_mask, n_attributes, n_values, remove_m
         masked_input = sender_input[mask==1].view(batch_size, -1)
     else:
         masked_input = sender_input*mask
+        if mask_by_last_value:
+            add_mask = torch.zeros((batch_size, n_attributes*n_values), device=sender_input.device)
+            idxs_to_one = (idxs_to_mask+1) * n_values - 1
+            add_mask = add_mask.scatter(dim=1, index=idxs_to_one, value=1)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/data.py]: Is this doing anything? Not sure I understand this part

File: egg/zoo/compo_vs_generalization/train.py [hunk 3]
```diff
@@ -163,16 +170,6 @@ def main(params):
...
 
-        train, test, validation, full_data = [
-            one_hotify(x, opts.n_attributes, opts.n_values)
-            for x in [train, test, validation, full_data]
-        ]
-
-        train, validation = ScaledDataset(train, 1), ScaledDataset(validation, 1)
-        test, full_data = ScaledDataset(test, 1), ScaledDataset(full_data, 1)
-        full_data_loader = DataLoader(full_data, batch_size=opts.batch_size)
-        print(" - done")
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py:174]: Any specific reason for removing this?

File: egg/zoo/compo_vs_generalization/train.py [hunk 4]
```diff
@@ -187,15 +184,22 @@ def main(params):
...
 
-        train, validation, test = [
-            one_hotify(x, opts.n_attributes, opts.n_values)
-            for x in [train, validation, test]
-        ]
+    if opts.masking_mode == 'dedicated_value':
+        # From this line onwards, n_values is added
+        # one value that represents masking the 
+        # corresponding attribute
+        opts.n_values += 1
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/train.py]: I am not sure if this is a safe thing to do. It then becomes a bit harder to tell which mode you're in within the game. At least  I think you should create a new variable like `tot_n_values` rather than overwriting the command line arguments.

File: launch.sh [hunk 5]
```diff
@@ -1,17 +1,13 @@
...
 
-# load conda environment
-source /shared/apps/anaconda3/etc/profile.d/conda.sh
-conda activate egg36
-
 # launch your computation
-python -m egg.zoo.compo_vs_generalization.train --n_epochs=10 --n_attributes=10 --n_values=10 --vocab_size=100 --max_len=40 --batch_size=5120 --sender_cell=lstm --receiver_cell=lstm --random_seed=38401 --data_scaler=1  --train_size=10000 --test_size=100 --validation_size=100 --curriculum
+# python -m egg.zoo.compo_vs_generalization.train --n_epochs=10 --n_attributes=10 --n_values=10 --vocab_size=100 --max_len=40 --batch_size=5120 --sender_cell=lstm --receiver_cell=lstm --random_seed=38401 --data_scaler=1  --train_size=10000 --test_size=100 --validation_size=100 --curriculum
 # python -m egg.zoo.compo_vs_generalization.data
+python test_cuda.py
```
Comment (pmichel31415) [inline launch.sh]: Might be worth switching to using egg/nest with json config files at some point (easier to track changes to experiments and launch sweeps)

File: plot_dataframe.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+    Each lines contains all the relevant hyperparameters, and the 
+    relevant data for a particular epoch.
+    """
+    df_lines = []
+    for line in file.readlines():
+        if line[:len('Namespace')] == 'Namespace':
+            params_dict = parse_params_line(line)
+            data = params_dict.copy()
+        if line[0] == '{':
```
Comment (pmichel31415) [inline plot_dataframe.py]: also `line.startswith("{")`

File: plot_dataframe.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            if i != 0: plt.show()
+            fig = plt.figure(figsize=(10, 5), dpi=120)
+            plt.subplots_adjust(wspace=0.5,
+                                hspace=0.5)
+        print(name)
+        ax = fig.add_subplot(2, 5, i%10+1)
+        title_indices = [RELEVANT_PARAMS.index(x)-1 for x in title_params]
+        title = '\n'.join([f"{title_params[i]}: {name[title_indices[i]]}" for i in range(len(title_params))]) 
+        ax.set_title(str(title).strip('{}'))
+        sb.lineplot(x=group['epoch'], y=group['acc_or'], ci='sd')
```
Comment (pmichel31415) [inline plot_dataframe.py]: You can also plot multiple curves on one plot with `sb.relplot`. It is used like this

```python
sb.relplot(
    kind="line",     # lineplot
    hue=XXX,         # use this to plot lines of different colors depending on parameter "XXX"  (note: you can use this with lineplot as well)
    row=YYY,         # different plot on each row depending on the value of param "YYY"
    col=ZZZ,         # different plot on each column depending on the value of param "ZZZ"
    **kwargs,         # otherwise same parameters as lineplot
)
```

**Review by pmichel31415 — CHANGES_REQUESTED** [review]
Nice job! Couple of things I didn't understand though, left a few comments

File: egg/zoo/compo_vs_generalization/train.py (hunk 3 above)
Comment (TheSpaceSheep) [reply egg/zoo/compo_vs_generalization/train.py:174]: I just did some code reorganizing here to prevent a useless repeat

File: egg/zoo/compo_vs_generalization/train.py (hunk 4 above)
Comment (TheSpaceSheep) [reply egg/zoo/compo_vs_generalization/train.py]: I felt so bad writing this but it was so convenient :D what you are suggesting seems indeed cleaner

File: launch.sh (hunk 5 above)
Comment (TheSpaceSheep) [reply launch.sh]: i'll untrack this file, as i'm indeed using sweeps know.

**Review by pmichel31415 — APPROVED** [review]
LGTM!
